src/lib/user_command_input/user_command_input.ml

```diff
@@ -0,0 +1,203 @@
+  {payload; sender; signature= sign_choice}
+
+let sign ~sender ~(user_command_payload : User_command_payload.t) = function
+  | Sign_choice.Signature signature ->
+      Option.value_map
+        ~default:(Deferred.return (Error "Invalid_signature"))
+        (User_command.create_with_signature_checked signature sender
+           user_command_payload)
+        ~f:Deferred.Result.return
+  | Keypair sender_kp ->
+      Deferred.Result.return (User_command.sign sender_kp user_command_payload)
+  | Hd_index hd_index ->
+      Secrets.Hardware_wallets.sign ~hd_index
+        ~public_key:(Public_key.decompress_exn sender)
+        ~user_command_payload
+
+let inferred_nonce ~f ~(sender : Public_key.Compressed.t) nonce_opt =
+  let open Result.Let_syntax in
+  let%bind txn_pool_or_account_nonce =
+    match Participating_state.active (f sender) |> Option.join with
+    | None ->
+        Error
+          "Couldn't infer nonce for transaction from specified `sender` since \
+           `sender` is not in the ledger or sent a transaction in transaction \
+           pool."
+    | Some nonce ->
+        Ok nonce
+  in
+  match nonce_opt with
+  | Some nonce ->
+      if Account_nonce.equal nonce txn_pool_or_account_nonce then Ok nonce
+      else
+        Error
+          (sprintf
+             !"Input nonce %s different from inferred nonce %s"
+             (Account_nonce.to_string nonce)
+             (Account_nonce.to_string txn_pool_or_account_nonce))
+  | None ->
+      Ok txn_pool_or_account_nonce
+
+let to_user_command ~result_cb ~infer_nonce ~logger uc_inputs :
+    (User_command.t list * (_ Or_error.t -> unit)) option Deferred.t =
+  let setup_user_command (client_input : t) :
+      User_command.t Deferred.Or_error.t =
+    (*create a user_command from input*)
+    let open Deferred.Or_error.Let_syntax in
+    let nonce_opt = client_input.payload.common.nonce in
+    let sender = client_input.sender in
+    let to_or_error : type a. (a, string) Result.t -> a Or_error.t =
+     fun e ->
+      let err_with_input str =
+        sprintf "Error creating user command: %s Error: %s"
+          (Yojson.Safe.to_string (to_yojson client_input))
+          str
+      in
+      Result.map_error e ~f:(Fn.compose Error.of_string err_with_input)
+    in
+    let%bind inferred_nonce =
+      inferred_nonce ~f:infer_nonce ~sender nonce_opt
+      |> to_or_error |> Deferred.return
+    in
+    let user_command_payload =
+      Payload.to_user_command_payload client_input.payload
+        ~nonce:inferred_nonce
+    in
+    let%map signed_user_command =
+      Deferred.map
+        (sign ~sender:client_input.sender ~user_command_payload
+           client_input.signature)
+        ~f:to_or_error
+    in
+    User_command.forget_check signed_user_command
+  in
+  let%map user_commands =
+    (*When batching multiple user commands, send all the user commands if they are valid or none if there is an error in one of them*)
+    let rec go acc = function
```

Comment:
This function reimplements `Deferred.Or_error.List.map`; it would be better to call the existing function instead.

Reply:
I was going for fold_until which is not implemented but `fold` would suffice with the new changes